conans/client/cmd/uploader.py

```diff
@@ -246,21 +263,48 @@ def upload_package(self, pref, prev_bundle, remote):
         self._output.debug(f"Upload {pref} in {duration} time")
 
 
-def compress_files(files, name, dest_dir, compresslevel=None, ref=None):
+def compress_files(files, name, dest_dir, compressformat=None, compresslevel=None, ref=None):
     t1 = time.time()
-    # FIXME, better write to disk sequentially and not keep tgz contents in memory
-    tgz_path = os.path.join(dest_dir, name)
-    if name in (PACKAGE_TGZ_NAME, EXPORT_SOURCES_TGZ_NAME) and len(files) > 100:
-        ref_name = f"{ref}:" or ""
+    tar_path = os.path.join(dest_dir, name)
+    if name in (PACKAGE_TGZ_NAME, PACKAGE_TZSTD_NAME, EXPORT_SOURCES_TGZ_NAME) and len(files) > 100:
+        ref_name = f"{ref}:" if ref else ""
         ConanOutput().info(f"Compressing {ref_name}{name}")
-    with set_dirty_context_manager(tgz_path), open(tgz_path, "wb") as tgz_handle:
-        tgz = gzopen_without_timestamps(name, mode="w", fileobj=tgz_handle,
-                                        compresslevel=compresslevel)
-        for filename, abs_path in sorted(files.items()):
-            # recursive is False in case it is a symlink to a folder
-            tgz.add(abs_path, filename, recursive=False)
-        tgz.close()
+
+    if compressformat == "zstd":
+        with open(tar_path, "wb") as tarfile_obj:
+            def reset_tarinfo(tarinfo):
+                """
+                Resets mtime in the tarinfo for consistency with
+                gzopen_without_timestamps()
+                """
+                tarinfo.mtime = 0
+                return tarinfo
+
+            # Only provide level if it was overridden by config.
+            zstd_kwargs = {}
+            if compresslevel is not None:
+                zstd_kwargs["level"] = compresslevel
+
+            dctx = zstandard.ZstdCompressor(**zstd_kwargs)
+
+            # Create a zstd stream writer so tarfile writes uncompressed data to
+            # the zstd stream writer, which in turn writes compressed data to the
+            # output tar.zst file.
+            with dctx.stream_writer(tarfile_obj) as stream_writer:
+                with tarfile.open(mode="w|", fileobj=stream_writer,
```

Comment:
You may want to increase the `bufsize` argument to around 512kB for both reading and writing. The default of 10kB is oddly chosen to best match TAR record sizes, but it's a total mis-fit for how storage stacks work.

ZSTD also works even better with larger block sizes, and since you don't have latency constraints, you can go wild.

It's also kind of stupid that the `tarfile.py` module doesn't use a pre-allocated buffers with `readinto` instead of `read` in the hot loop, but that's not something you can fix here:
https://github.com/python/cpython/blob/3.12/Lib/tarfile.py#L251C19-L251C23

Reply:
Btw.: I suspect this is also part of the reason why the Gzip path is so slow / GIL heavy.

Reply:
I tested decompression with various `bufsize` values and am finding that the sweet spot is right around 32KB. I'll run more tests to find the best value but my first test for a 7GB package with 16000 files showed this:
```
Decompress+extract for bufsize 10KB took 107.5195472240448 seconds
Decompress+extract for bufsize 16KB took 109.06598091125488 seconds
Decompress+extract for bufsize 32KB took 104.0440125465393 seconds
Decompress+extract for bufsize 64KB took 111.28909516334534 seconds
Decompress+extract for bufsize 128KB took 103.44610142707825 seconds
Decompress+extract for bufsize 256KB took 126.36465835571289 seconds
Decompress+extract for bufsize 512KB took 128.68103170394897 seconds
Decompress+extract for bufsize 1024KB took 141.1732087135315 seconds
Decompress+extract for bufsize 2048KB took 290.1585919857025 seconds
```